pkg/controller/v1alpha1/localmodel/controller.go

```diff
@@ -0,0 +1,658 @@
+}
+
+var (
+	ownerKey         = ".metadata.controller"
+	localModelKey    = ".localmodel"
+	apiGVStr         = v1alpha1api.SchemeGroupVersion.String()
+	modelCacheCRName = "ClusterLocalModel"
+	finalizerName    = "localmodel.kserve.io/finalizer"
+	defaultImage     = "kserve/storage-initializer:latest" // Updated in reconcile()
+)
+
+// The localmodel is being deleted
+func (c *LocalModelReconciler) deleteModelFromNodes(localModel *v1alpha1.ClusterLocalModel, jobNamespace string) (ctrl.Result, error) {
+	// finalizer does not exists, nothing to do here!
+	if !utils.Includes(localModel.ObjectMeta.Finalizers, finalizerName) {
+		return ctrl.Result{}, nil
+	}
+
+	// Check job status and decide whether to delete finalizer
+	c.Log.Info("deleting, finalizer exists")
+
+	// Todo: Prevent deletion if there are isvcs using this localmodel
+
+	allDone := true
+	for node := range localModel.Status.NodeStatus {
+		jobName := localModel.Name + "-" + node + "-delete"
+
+		job, err := c.launchDeletionJob(jobName, jobNamespace, localModel, localModel.Spec.SourceModelUri, localModel.Name, node)
+		if err != nil {
+			c.Log.Error(err, "Deletion Job err", "name", jobName)
+			return ctrl.Result{}, err
+		}
+		switch {
+		case job.Status.Succeeded > 0:
+			c.Log.Info("Deletion Job succeeded", "name", jobName)
+			localModel.Status.NodeStatus[node] = v1alpha1api.NodeDeleted
+		case job.Status.Failed > 0:
+			allDone = false
+			localModel.Status.NodeStatus[node] = v1alpha1api.NodeDeletionError
+		default:
+			allDone = false
+			localModel.Status.NodeStatus[node] = v1alpha1api.NodeDeleting
+		}
+	}
+	if allDone {
+		patch := client.MergeFrom(localModel.DeepCopy())
+		// remove our finalizer from the list and update it.
+		localModel.ObjectMeta.Finalizers = utils.RemoveString(localModel.ObjectMeta.Finalizers, finalizerName)
+		if err := c.Patch(context.Background(), localModel, patch); err != nil {
+			c.Log.Error(err, "Cannot remove finalizer", "model name", localModel.Name)
+			return ctrl.Result{}, err
+		}
+	}
+	if err := c.Status().Update(context.Background(), localModel); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Stop reconciliation as the item is being deleted
+	return ctrl.Result{}, nil
+}
+
+// Creates a PV and set the localModel as its controller
+func (c *LocalModelReconciler) createPV(spec v1.PersistentVolume, localModel *v1alpha1.ClusterLocalModel) error {
+	persistentVolumes := c.Clientset.CoreV1().PersistentVolumes()
+	if _, err := persistentVolumes.Get(context.TODO(), spec.Name, metav1.GetOptions{}); err != nil {
+		if !apierr.IsNotFound(err) {
+			c.Log.Error(err, "Failed to get PV")
+			return err
+		}
+		c.Log.Info("Create PV", "name", spec.Name)
+		if err := controllerutil.SetControllerReference(localModel, &spec, c.Scheme); err != nil {
+			c.Log.Error(err, "Failed to set controller reference")
+			return err
+		}
+		if _, err := persistentVolumes.Create(context.TODO(), &spec, metav1.CreateOptions{}); err != nil {
```

Comment:
How do we create `storageClass`? I think we need an option to create that which can be set on the PV.

Reply:
We have discussed this before. ClusterLocalModel does not support custom storageclass. Namespace scoped CR can support custom storageclass which would be a different controller.